cmd/snap-bootstrap/cmd_scan_disk.go

```diff
@@ -0,0 +1,305 @@
+ * for implementation.
+ *
+ * Note that symlink /dev/disk/snapd/disk can be expected by
+ * snap-bootstrap. In that case, snap-initramfs-mounts.service should
+ * have:
+ *
+ * ```
+ * BindsTo=dev-disk-snapd--disk.device
+ * After=dev-disk-snapd-disk.device
+ * ```
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/cmd/snap-bootstrap/blkid"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil/kcmdline"
+)
+
+func init() {
+	const (
+		short = "Verify that a disk is the booting disk"
+		long  = "This tool is expected to be called from udev"
+	)
+
+	addCommandBuilder(func(parser *flags.Parser) {
+		if _, err := parser.AddCommand("scan-disk", short, long, &cmdScanDisk{}); err != nil {
+			panic(err)
+		}
+	})
+}
+
+type cmdScanDisk struct{}
+
+func (c *cmdScanDisk) Execute([]string) error {
+	return ScanDisk(os.Stdout)
+}
+
+type Partition struct {
+	Name string
+	UUID string
+}
+
+func isGpt(probe blkid.AbstractBlkidProbe) bool {
+	pttype, err := probe.LookupValue("PTTYPE")
+	if err != nil {
+		return false
+	}
+	return pttype == "gpt"
+}
+
+func probePartitions(node string) ([]Partition, error) {
+	probe, err := blkid.NewProbeFromFilename(node)
+	if err != nil {
+		return nil, err
+	}
+	defer probe.Close()
+
+	probe.EnablePartitions(true)
+	probe.SetPartitionsFlags(blkid.BLKID_PARTS_ENTRY_DETAILS)
+	probe.EnableSuperblocks(true)
+
+	if err := probe.DoSafeprobe(); err != nil {
+		return nil, err
+	}
+
+	if !isGpt(probe) {
+		return nil, nil
+	}
+
+	partitions, err := probe.GetPartitions()
+	if partitions == nil {
+		return nil, err
+	}
+
+	ret := make([]Partition, 0)
+	for _, partition := range partitions.GetPartitions() {
+		label := partition.GetName()
+		uuid := partition.GetUUID()
+		ret = append(ret, Partition{label, uuid})
+	}
+
+	return ret, nil
+}
+
+func samePath(a, b string) (bool, error) {
+	aSt, err := os.Stat(a)
+	if err != nil {
+		return false, err
+	}
+	bSt, err := os.Stat(b)
+	if err != nil {
+		return false, err
+	}
+	return os.SameFile(aSt, bSt), nil
+}
+
+
+func scanDiskNodeFallback(output io.Writer, node string) error {
+	var fallbackPartition string
+
+	partitions, err := probePartitions(node)
+	if err != nil {
+		return fmt.Errorf("cannot get partitions: %s\n", err)
+	}
+	/*
+	 * If LoaderDevicePartUUID was not set, it is probably because
+	 * we did not boot with UEFI. In that case we try to detect
+	 * disk with partition labels.
+	 */
+
+	mode, _, err := boot.ModeAndRecoverySystemFromKernelCommandLine()
+	if err != nil {
+		return err
+	}
+	switch mode {
+	case "recover":
+		fallbackPartition = "ubuntu-seed"
+	case "install":
+		fallbackPartition = "ubuntu-seed"
+	case "factory-reset":
+		fallbackPartition = "ubuntu-seed"
+	case "run":
+		fallbackPartition = "ubuntu-boot"
+	case "cloudimg-rootfs":
+		fallbackPartition = "ubuntu-boot"
```

Comment:
These lines could we removed as we now know fallback part is not relevant for CVM.